cpp/src/arrow/array/builder_binary.h

```diff
@@ -136,70 +276,119 @@ class ARROW_EXPORT BinaryBuilder : public ArrayBuilder {
+    if (i == (length_ - 1)) {
+      *out_length = static_cast<offset_type>(value_data_builder_.length()) - offset;
+    } else {
+      *out_length = offsets[i + 1] - offset;
+    }
+    return value_data_builder_.data() + offset;
+  }
 
   /// Temporary access to a value.
   ///
   /// This view becomes invalid on the next modifying operation.
-  util::string_view GetView(int64_t i) const;
+  util::string_view GetView(int64_t i) const {
+    const offset_type* offsets = offsets_builder_.data();
+    const auto offset = offsets[i];
+    offset_type value_length;
+    if (i == (length_ - 1)) {
+      value_length = static_cast<offset_type>(value_data_builder_.length()) - offset;
+    } else {
+      value_length = offsets[i + 1] - offset;
+    }
+    return util::string_view(
+        reinterpret_cast<const char*>(value_data_builder_.data() + offset), value_length);
+  }
 
  protected:
-  TypedBufferBuilder<int32_t> offsets_builder_;
+  TypedBufferBuilder<offset_type> offsets_builder_;
   TypedBufferBuilder<uint8_t> value_data_builder_;
 
-  Status AppendOverflow(int64_t num_bytes);
+  Status AppendOverflow(int64_t num_bytes) {
+    return Status::CapacityError("array cannot contain more than ", memory_limit(),
+                                 " bytes, have ", num_bytes);
+  }
 
   Status AppendNextOffset() {
     const int64_t num_bytes = value_data_builder_.length();
-    if (ARROW_PREDICT_FALSE(num_bytes > kBinaryMemoryLimit)) {
+    if (ARROW_PREDICT_FALSE(num_bytes > memory_limit())) {
       return AppendOverflow(num_bytes);
     }
-    return offsets_builder_.Append(static_cast<int32_t>(num_bytes));
+    return offsets_builder_.Append(static_cast<offset_type>(num_bytes));
   }
 
   void UnsafeAppendNextOffset() {
     const int64_t num_bytes = value_data_builder_.length();
-    offsets_builder_.UnsafeAppend(static_cast<int32_t>(num_bytes));
+    offsets_builder_.UnsafeAppend(static_cast<offset_type>(num_bytes));
+  }
+
+  // Cannot make this a static attribute because of linking issues
+  static constexpr int64_t memory_limit() {
+    return std::numeric_limits<offset_type>::max() - 1;
   }
 };
 
+/// \class BinaryBuilder
+/// \brief Builder class for variable-length binary data
+class ARROW_EXPORT BinaryBuilder : public BaseBinaryBuilder<BinaryType> {
+ public:
+  explicit BinaryBuilder(MemoryPool* pool ARROW_MEMORY_POOL_DEFAULT);
+
+  /// \cond FALSE
+  using ArrayBuilder::Finish;
+  /// \endcond
+
+  Status Finish(std::shared_ptr<BinaryArray>* out) { return FinishTyped(out); }
+
+ protected:
+  using BaseBinaryBuilder::BaseBinaryBuilder;
+};
+
 /// \class StringBuilder
 /// \brief Builder class for UTF8 strings
 class ARROW_EXPORT StringBuilder : public BinaryBuilder {
  public:
   using BinaryBuilder::BinaryBuilder;
   explicit StringBuilder(MemoryPool* pool ARROW_MEMORY_POOL_DEFAULT);
 
-  using BinaryBuilder::Append;
-  using BinaryBuilder::Reset;
-  using BinaryBuilder::UnsafeAppend;
+  /// \cond FALSE
```

Comment:
what is this for?

Reply:
To avoid Doxygen barfing on different overloads of a method IIRC :-/